text/3334-niche.md

```diff
@@ -0,0 +1,322 @@
+common type of niche, while leaving room for future extension.
+
+Defining a niche mechanism allows libraries to build arbitrary types containing
+niches, and simplifies handling of space-efficient data structures in Rust
+without manual bit-twiddling.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a structure containing exactly one field of a non-zero-sized type
+(non-ZST), you can attach a `niche` attribute on it to declare a specific value
+or range of values for that field as invalid. This promises the compiler that
+you will never store those values in that field, which allows the compiler to
+use those in-memory representations for different purposes, such as the
+representation of `None` in a containing `Option`.
+
+```rust
+use std::mem::size_of;
+
+#[niche(value = 42)]
+struct MeaninglessNumber(u64);
+
+assert_eq!(size_of::<MeaninglessNumber>(), 8);
+assert_eq!(size_of::<Option<MeaninglessNumber>>(), 8);
+
+#[niche(range = 2..)]
+struct Bit(u8);
+
+assert_eq!(size_of::<Bit>(), 1);
+assert_eq!(size_of::<Option<Option<Option<Bit>>>>(), 1);
+```
+
+Constructing a structure with a niche value, or writing to the non-ZST field of
+such a structure, or obtaining a mutable reference to such a field, requires
+`unsafe` code. Causing a type with a niche to contain an invalid value (whther
+by construction, writing, or transmuting) results in undefined behavior.
+
+If a type `T` contains only a single niche value, `Option<T>` (and other enums
+isomorphic to it, with one variant containing `T` and one nullary variant) will
+use that value to represent `None` (the nullary variant). If such a `T` is
+additionally `repr(transparent)` or `repr(C)` or otherwise permitted in FFI,
+`Option<T>` will likewise be permitted in FFI, with the niche value mapping
+bidirectionally to `None` across the FFI boundary.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The attribute `#[niche]` may only appear on a struct declaration. The struct
+must contain exactly one field of a non-zero-sized type (non-ZST). The struct
+may contain zero or more ZST fields, such as `PhantomData`.
+
+Declaring a niche on any item other than a struct declaration results in an
+error.
+
+Declaring multiple `niche` attributes on a single item, or multiple key-value
+pairs within a single `niche` attribute, results in an error.
+
+Declaring a niche on a struct containing more or less than one non-zero-sized
+field results in an error.
+
+Declaring a range niche with an empty range (e.g. `0..0`) results in a
+warn-by-default lint. As with many lints, this lint should be automatically
+suppressed for code expanded from a macro.
+
+Declaring a range niche with an invalid range (e.g. `5..0`) results in an
+error.
+
+Declaring a niche using a negative value or a negative range endpoint results
+in an error. The text of the error should suggest the appropriate unsigned
```

Comment:
This was surprising. You have a little bit of explanation later, but may be worth a few more words somewhere. It's mentioned that the compiler isn't smart enough, but that feels light to me.

Reply:
Done, but let me know if it needs further improvement.